scipy/spatial/src/distance_impl.h

```diff
@@ -309,6 +310,39 @@ jaccard_distance_char(const char *u, const char *v, const npy_intp n)
     return (double)num / denom;
 }
 
+
+static NPY_INLINE double
+jensenshannon_distance_double(const double *p, const double *q, const npy_intp n)
+{
+    npy_intp i;
+    double s = 0.0;
+    double p_sum = 0.0;
+    double q_sum = 0.0;
+
+    for (i = 0; i < n; ++i) {
+        if (p[i] < 0.0 || q[i] < 0.0)
+            return HUGE_VAL;
+        p_sum += p[i];
+        q_sum += q[i];
+    }
+
+    if (p_sum == 0.0 || q_sum == 0.0)
+        return HUGE_VAL;
+
+    for (i = 0; i < n; ++i) {
+        const double p_i = p[i] / p_sum;
```

Comment:
I'm a little confused here -- you've declared a few variables to be of a constant type, and yet they change value with each iteration of the loop. Maybe they should just be defined as double type at the start of the function?

Reply:
The `const` keyword means constant until the variable goes out of scope, which is until the end of each loop in this case. The C implementation of `seuclidean` also does something similar, i.e.,

```
for (i = 0; i < n; ++i) {
    const double d = u[i] - v[i];
    s += (d * d) / var[i];
}
```

Reply:
I wonder if this actually performs better with modern compilers vs. declaring all variable types at the start of the function. This seems to mostly be considered a matter of style in the SO posts I've browsed.

Reply:
I'm guessing that modern compilers can optimize away the difference, especially when there are no pointers involved and the loop is only six lines long. The main function of `const`  keyword here is really to prevent the coder (me) from doing something silly with values that shouldn't be changed in each iteration of the loop.